### ISSUE TITLE
Enabling mTLS Auto Causes Kubernetes Worker Nodes to become unresponsive

### ISSUE THREAD

**Issue by crhuber** [istio/istio#20797]
**Bug description**
When `disabling` or `enabling` auto mTLS via helm chart on Istio 1.4.3 causes Kubernetes worker nodes to go into a `NotReady` State. (We have the Istio Control Plane nodes in a different worker group so they are unaffected)

Set via Helm:
```
  mtls:
    auto: true or false
```

As soon as this is set to `true` or `false`, it causes roughly 25 worker nodes of our ~75 node cluster to go into a `NotReady` state.  When logging into the worker nodes the CPU is usually pegged at 100% with the top processes alternating between: 
`kswapd0`, `iptables-restore`, `dockerd`, `kubelet`, `kube-proxy` and a few others.
 
![image](https://user-images.githubusercontent.com/5857391/73675548-e64ab480-46b2-11ea-978b-77ee2e09304b.png)
 
The nodes have high CPU load and are unresponsive to ssh or commandline. But eventually the problem resolves itself after about 90 mins (as seen above) without any intervention. I was able to run top from a few nodes but it takes about 2 mins just to run top. Here is the output at various times over the 90mins it occurs. 

**Top Output**

```
  PID USER  PR  NI    VIRT    RES    SHR S  %CPU %MEM     TIME+ COMMAND
  457       20   0       0      0      0 R  77.7  0.0  59:20.73 kswapd0
 4016       20   0 2009228 199540   4704 S  72.4  2.6 872:06.84 dockerd
16648       20   0 2736864 209076   9828 S  27.9  2.7   1490:37 kubelet
11088       20   0  891108  36576   8244 S  16.6  0.5   0:42.68 node
26106       20   0  376160  86684  77080 S  12.0  1.1   6:35.80 searchd
10179       20   0  123448  10708      0 S  10.6  0.1   0:21.63 
```

```
 PID USER  PR  NI    VIRT    RES    SHR S  %CPU %MEM     TIME+ COMMAND
  457       20   0       0      0      0 R  96.4  0.0 203:05.85 kswapd0
 5476       20   0   31084   1664   1032 R  67.5  0.0   0:28.16 iptables-restor
20458       20   0  212372  35124      0 S  41.7  0.4   7:19.84 occollector_lin
 4753       20   0 2372192  87512  12872 S  40.7  1.1   0:31.86 kubelet
 6800       20   0  923332 103224      0 S  24.8  1.3   1:07.08 gunicorn
```

```
14:03:22 up 5 days,  5:39,  1 user,  load average: 78.02, 68.19, 47.99
Tasks: 362 total,  14 running, 292 sleeping,   0 stopped,   2 zombie
%Cpu(s):  1.8 us, 98.1 sy,  0.0 ni,  0.0 id,  0.0 wa,  0.0 hi,  0.1 si,  0.0 st
KiB Mem :  7807644 total,   101028 free,  7490976 used,   215640 buff/cache
KiB Swap:        0 total,        0 free,        0 used.    58976 avail Mem

  PID USER  PR  NI    VIRT    RES    SHR S  %CPU %MEM     TIME+ COMMAND
  457       20   0       0      0      0 R  99.0  0.0  18:49.19 kswapd0
 5204       20   0   30904    520     80 R  50.8  0.0   0:06.10 iptables-restor
 8168       20   0 1090180   8852      0 R  49.5  0.1   6:50.58 service1
24382       20   0 1165232   6624      0 S  31.2  0.1   3:42.95 ad
 5448 xyz   20   0  191320  71212    268 S  23.9  0.9   0:20.18 celery
22119       20   0  113968   2444      0 S  23.1  0.0   0:38.13 php-fpm_exporte
27674       20   0 1096372   7028      0 S  22.1  0.1   1:25.16 service2
 7882 1337  20   0  368780 232120   3732 R  16.5  3.0  24:25.70 envoy
```

```
top - 14:23:17 up 6 days, 22:49,  1 user,  load average: 59.34, 63.27, 100.72
Tasks: 395 total,  20 running, 325 sleeping,   0 stopped,   1 zombie
%Cpu(s):  8.2 us, 91.7 sy,  0.0 ni,  0.0 id,  0.0 wa,  0.0 hi,  0.1 si,  0.0 st
KiB Mem :  7864988 total,   103196 free,  7451664 used,   310128 buff/cache
KiB Swap:        0 total,        0 free,        0 used.    97920 avail Mem

  PID USER  PR  NI    VIRT    RES    SHR S  %CPU %MEM     TIME+ COMMAND
  457       20   0       0      0      0 R  85.8  0.0 117:22.18 kswapd0
23166       20   0  140144  36944      0 R  39.7  0.5   0:01.25 kube-proxy

```

```
14:30:48 up 6 days, 22:57,  1 user,  load average: 76.30, 79.92, 93.75
Tasks: 395 total,  13 running, 332 sleeping,   0 stopped,   0 zombie
%Cpu(s):  3.8 us, 96.0 sy,  0.0 ni,  0.0 id,  0.1 wa,  0.0 hi,  0.1 si,  0.0 st
KiB Mem :  7864988 total,   104552 free,  7460892 used,   299544 buff/cache
KiB Swap:        0 total,        0 free,        0 used.    94296 avail Mem

  PID USER  PR  NI    VIRT    RES    SHR S  %CPU %MEM     TIME+ COMMAND
23565       20   0   47820   8612   1824 R  94.1  0.1   0:10.50 iptables-restor
  457       20   0       0      0      0 R  87.9  0.0 123:48.38 kswapd0
```

```
14:56:42 up 5 days,  5:41,  1 user,  load average: 42.38, 50.48, 58.92
Tasks: 397 total,  12 running, 336 sleeping,   0 stopped,   0 zombie
%Cpu(s):  4.2 us, 95.6 sy,  0.0 ni,  0.1 id,  0.0 wa,  0.0 hi,  0.1 si,  0.0 st
KiB Mem :  7807644 total,   103436 free,  7324420 used,   379788 buff/cache
KiB Swap:        0 total,        0 free,        0 used.   152248 avail Mem

  PID USER  PR  NI    VIRT    RES    SHR S  %CPU %MEM     TIME+ COMMAND
  458       20   0       0      0      0 R  70.2  0.0  63:19.21 kswapd0
 8280       20   0  140080  36368      0 R  51.5  0.5   0:02.71 kube-proxy
```

Here is the  node load average that happens over that time period. The lines are broken because the metrics couldnt be retrieved during high load 

![image](https://user-images.githubusercontent.com/5857391/73676023-d97a9080-46b3-11ea-8266-c2fdccb34359.png)

**Pilot CPU** (they were unaffected)
![image](https://user-images.githubusercontent.com/5857391/73676068-f44d0500-46b3-11ea-9a12-4cc054d9c31f.png)

**Pilot Pushes** 
![image](https://user-images.githubusercontent.com/5857391/73676089-0169f400-46b4-11ea-950b-4a427e978107.png)
I dont see any huge amount of pushes during this time

**Pilot Errors**
![image](https://user-images.githubusercontent.com/5857391/73676123-19417800-46b4-11ea-8da8-959e35797655.png)
I think maybe due to Pilot unable to reach the envoys because the nodes were unresponsive??

**Pilot Conflicts**
![image](https://user-images.githubusercontent.com/5857391/73676171-38400a00-46b4-11ea-8aa1-cd2092a65a22.png)
No more than usual

**Global Envoy Memory Usage**
![image](https://user-images.githubusercontent.com/5857391/73676225-4ee66100-46b4-11ea-971b-d5b125bb4e68.png)
It spikes up during this time. 

Sample of One Node's **CPU Load** Right at the time Issue Occurs
![image](https://user-images.githubusercontent.com/5857391/73676282-6aea0280-46b4-11ea-8c7b-c7526f283360.png)

Sample of One Nodes **Memory Available Bytes** at time Issue Occurs
![image](https://user-images.githubusercontent.com/5857391/73676374-8ce38500-46b4-11ea-9fca-41717e144a10.png)
Notice available memory drops to 0
 

[ ] Configuration Infrastructure
[ ] Docs
[ ] Installation
[x ] Networking
[ ] Performance and Scalability
[ ] Policies and Telemetry
[ ] Security
[ ] Test and Release
[ ] User Experience
[ ] Developer Infrastructure

**Expected behavior**
Normal CPU load

**Steps to reproduce the bug**

Set 
```
  mtls:
    auto: true 
```
or Set

```
  mtls:
    auto: false 
```

**Version (include the output of `istioctl version --remote` and `kubectl version` and `helm version` if you used Helm)**

client version: 1.4.2
citadel version: 1.4.3
citadel version: 1.4.3
galley version: 1.4.3
galley version: 1.4.3
ingressgateway version: 1.4.3
ingressgateway version: 1.4.3
pilot version: 1.4.3
pilot version: 1.4.3
pilot version: 1.4.3
pilot version: 1.4.3
pilot version: 1.4.3
pilot version: 1.4.3
policy version: 1.4.3
policy version: 1.4.3
sidecar-injector version: 1.4.3
sidecar-injector version: 1.4.3
telemetry version: 1.4.3
telemetry version: 1.4.3
data plane version: 1.4.1 (3 proxies), 1.4.3 (951 proxies), 1.4.0 (33 proxies)

Kube:
Server Version: version.Info{Major:"1", Minor:"14+", GitVersion:"v1.14.9-eks-c0eccc"}

**How was Istio installed?**
Helm

**Environment where bug was observed (cloud vendor, OS, etc)**

Amazon EKS.

**Comment by howardjohn** [comment]
cc @incfly @diemtvu 

This is pretty weird. Can you reproduce in any way other than auto-mtls? Like changing MeshPolicy or some other field in mesh config?

**Comment by diemtvu** [comment]
Can you change the DR to ISTIO_MUTUAL globally while keep every thing the same. This will preempt the auto-mtls logic, but the end effect on data plane should be the same (all traffic use mTLS). If you see the same pattern, then it could be because of the extra cost due to handling mTLS.

Below is the example DR for your reference.

```
apiVersion: "networking.istio.io/v1alpha3"
kind: "DestinationRule"
metadata:
  name: "default"
spec:
  host: "*.local"
  trafficPolicy:
    tls:
      mode: ISTIO_MUTUAL
```

**Comment by crhuber** [comment]
@diemtvu most of the services already have a DestinationRule already without the `tls` key in it.  From what I understand creating a DR as per your example, wouln't actually reconfigure anything on the envoy side because I think the most specific rule applies, which, in this case would be the already existing one. 

I tried a small test to apply a DR to a namespace as such:

```
apiVersion: "networking.istio.io/v1alpha3"
kind: "DestinationRule"
metadata:
  name: "default"
  namespace: ns1
spec:
  host: "*.local"
  trafficPolicy:
    tls:
      mode: ISTIO_MUTUAL
```

however the clusters a client sidecar were not changed, I believe because this DR already exists

```
kind: DestinationRule
apiVersion: networking.istio.io/v1alpha3
metadata:
  name: sampleapp
  namespace: ns1
spec:
  host: app.ns1.svc.cluster.local
  trafficPolicy:
    loadBalancer:
      simple: ROUND_ROBIN
    outlierDetection:
      consecutiveErrors: 5 
      interval: 10s 
      baseEjectionTime: 30s 
      maxEjectionPercent: 20
    connectionPool:
      http:
        http1MaxPendingRequests: 1
      tcp:
        maxConnections: 10
```

**Comment by crhuber** [comment]
@howardjohn I havent been able to reproduce other than by setting auto: false or auto: true. Ive tried it 3 times to confirm it wasnt a random occurance

**Comment by howardjohn** [comment]
That is very weird. I am a bit confused by your top output though, you said those processes are using 100% cpu but looks like the worst is `2.7   1490:37 kubelet` at 2.7% cpu? Am I reading this wrong?

**Comment by incfly** [comment]
I just tried our implementation in 1.4.3, that when `DestinationRule.ISTIO_MUTUAL` exists, our cluster.tls configuration is the same. when i flip the mesh config true/false, while I do see the push happened, but the tls config of the cluster remains the same...

however, i saw your comment in https://github.com/istio/istio/issues/20797#issuecomment-581840999

are you saying many of your DestinationRule is as the second example

``` 
kind: DestinationRule
apiVersion: networking.istio.io/v1alpha3
metadata:
  name: sampleapp
  namespace: ns1
spec:
  host: app.ns1.svc.cluster.local
  trafficPolicy:
    loadBalancer:
      simple: ROUND_ROBIN
    outlierDetection:
      consecutiveErrors: 5 
      interval: 10s 
      baseEjectionTime: 30s 
      maxEjectionPercent: 20
    connectionPool:
      http:
        http1MaxPendingRequests: 1
      tcp:
        maxConnections: 10
```

basically, exists, but not TLS config specified? in that case, auto mTLS will kick in, flipping the mesh config we shall see actual client side envoy outbound cluster configuration change. (you can verify that by doing `istioctl pc cluster <pod-id> -ojson` and look at any cluster config. should see it changed from `tls -> transport_socket_match` field)

could you confirm your mesh's DR is falling into that case? (exists but no TLS sections).

**Comment by crhuber** [comment]
> That is very weird. I am a bit confused by your top output though, you said those processes are using 100% cpu but looks like the worst is `2.7 1490:37 kubelet` at 2.7% cpu? Am I reading this wrong?

@howardjohn 
There was a spacing issue in the output. I fixed it now

**Comment by crhuber** [comment]
@incfly . Yes we have DR's as per my example for many services and they do not have any tls config in them. When `mtls: auto: true / false` happens I can confirm the envoy cluster configuration does change and and that mTLS does indeed work after that. However the issue is that by enabling it or disabling it kills Kubernetes nodes. 
Bear in mind we have 900 connected envoys

**Comment by howardjohn** [comment]
Not an expert here, but the kswap0 stuff makes me think looking into the memory usage may be important, and the CPU is a just a result of that. If we are consuming all the memory on the node it could explain all the other processes using more CPU as well?

**Comment by incfly** [comment]
I once was told by @lambdai that memory overhead of cluster is none trivial. global flipping auto mtls flag in mesh config makes all the outbound cluster within the mesh changed. I don't know the details of how envoy updating cluster when the config changed, in terms of numbers, but flag update -> entire mesh sidecar proxy, all the outbound clusters updated, -> mem footprint surge, this sequencing seems the culprint

i'm a bit surprised this cause such significant effect though. doesn't this mean any feature that can cause mesh-wide outbound cluster can cause similar issue?

1.5 istiod allows multi environment, multi control plane. can users control the schedule of the rollout? i.e first 1 hour, 10% of the envoys using the 1.5 istio control plane, 2 hr, 20%, etc.

alternatively, are there some internal debouncing/throttling knob can help us during this global flag change? @howardjohn wdty?

**Comment by howardjohn** [comment]
It will probably be more like 1.6 before multi env is fully ready. But once we have that, yeah you can easily do a canary type rollout of this. We can also potentially do some sort of throttling/debouncing but we need to figure out what is going on to see what is the correct mechanism to fix it.

**Comment by incfly** [comment]
> figure out what is going on

will it be enough if we can confirm the memory bump indeed comes from the CDS update and get numbers?

**Comment by howardjohn** [comment]
that part I think is safe to assume but typically a jump in memory doesn't crash the whole cluster like this?

**Comment by incfly** [comment]
i have similar feeling as well. @howardjohn  are you aware of any prior feature roll-out that can trigger entire mesh outbound cluster being updated? if we don't this might just be the first time we hit it.

at the same time, some calculation might help to understand why crash the cluster. i can think of things like,

- the number of the services in the mesh
- the CPU/memory capacity of the nodes
- average number of the Istio injected pods on each nodes.

@crhuber could you provide some numbers of above? thanks!

**Comment by crhuber** [comment]
@incfly 
 
* 750 services
* 985 istio-proxy
* 70 nodes (4 VCPU, 8 GB)
* 13-18 avg istio-proxy per node

**Comment by howardjohn** [comment]
Can you look more into the memory? specifically the memory usage relative to the total memory on the node? Are we just hitting the ceiling on memory and things falling apart? 

I personally do most testing on GCP n1-standard nodes which have 2x the cpu:memory as yours so that would explain why you see that and I haven't

**Comment by incfly** [comment]
One more thing, I learned envoy `--concurrency` flag affects the memory consumption as well. This determines the number of the work thread envoy spawns up, and each core has its own thread local storage.

you can get this by looking at any envoy pod yaml, `kubectl get po -oyaml <pod-id>`, and peek the `--concurrency` flag value from the pod spec.

**Comment by crhuber** [comment]
@incfly    
`--concurrency = "2"`

**Comment by crhuber** [comment]
@howardjohn I looked into some of the memory numbers:

- On average the nodes have 2.5GB free
- On average the "bottom" 25 nodes have only 1.6GB free 
- On average each node runs 25 istio-proxy 
- On average each istio-proxy consumes 150mb memory
- Enabling mTLS istio-proxy memory shot up to 210mb 
-  60mb extra required * 25 istio-proxy per node = 1.5GB required
- 1.6GB Free - 1.5GB = kswapd0 = 💣

**Comment by incfly** [comment]
I deployed service graph from istio [perf test suite](https://github.com/istio/tools/blob/master/perf/load/setup_large_test.sh), with 10 namespaces, each ns has 20 services, 40 pods, so 400+ pods in total. I delete the default `sidecar` and `destinationrule`, which are part of the initial test setup. removing sidecar bumps up per proxy mem consumption from 35 -> 55 mb

and then i flip the mesh config auto mTLS option from false to true, then i see the mem/proxy changed to 70 mb. And never drops back.

![cluster-mem-no-drop](https://user-images.githubusercontent.com/1661831/73881205-c4694300-4814-11ea-9e57-3db88ee231b5.png)

so compared with your observation

- in your plot, after flipping the mesh config auto mTLS option, after a while being stablized, mem consumption per proxy dropped, which I didn't see in my setup. weird
- the actual per proxy increasing usage could depends on the other Istio features we used, also the traffic, I assume. I saw a 55 -> 70 MB bump up, but your case, it's 160 -> 210 MB. how much capactiy reservation might depends on the workload size/traffic.

**Comment by incfly** [comment]
I tried one more round of measuring to see the impact of flipping this flag, still failed to reproduce the the memory consumption as huge as reported in this issue.

the order of the sequence in the graph, per proxy memory.

- start with DestaintionRule ISTIO_MUTUAL, with sidecar resource limiting scope. 30MB / proxy
- 13:24 pm, remove the sidecar scoping CRD, 7:40; 55 MB/proxy
- remove DR change to plaintext. goes up a bit first, and then goes down to 55 MB again
- 13:55 pm, start flip auto mTLS. 65 MB 
- memory does not drop. weird.
- restart all proxy, 15:00 pm. now drops to 58 MB.
- 16:37, flip the automTLS to false, see bump up to 60+ MB.

![cluster-mem-no-drop](https://user-images.githubusercontent.com/1661831/73991127-1a66e500-4900-11ea-99b3-59071d77f47a.png)


https://snapshot.raintank.io/dashboard/snapshot/CwT637hQETSeNku63v950ampPtu27Quy?orgId=2&fullscreen&panelId=902

**Comment by incfly** [comment]
Also given the Envoy mem consumption depends on other factors, what Envoy features are in used, the number of the hosts in each Envoy cluster, I don't think there's one number (X %)  fitting for all cases.

The mitigation I can think: offering a per namespace control to allow customer roll out gradually by label namespace. We implement this in Pilot/Istiod, the namespace labeling takes higher priority than mesh config. And we can document this knob correspondingly.

**Comment by crhuber** [comment]
@incfly @howardjohn  thanks for your help on this one guys. 

We were able to safely roll the feature configuring each worker node to have more memory headroom. We did this by scaling out the number of nodes and then configuring each node to have a lower `maxPods` setting in k8s. One this was done, we were able to turn on mtls without issue. 

closing
